### PR TITLE
Adding a refresh listener to a recovering shard should be a noop

### PR DESCRIPTION
**Pull request by bleskes**
When `refresh=wait_for` is set on an indexing request, we register a listener on the shards that are call during the next refresh. During the recover translog phase, when the engine is open, we have a window of time when indexing operations succeed and they can add their listeners. Those listeners will only be called when the recovery finishes as we do not refresh during recoveries (unless the indexing buffer is full). Next to being a bad user experience, it can also cause deadlocks with an ongoing peer recovery that may wait for those operations to mark the replica in sync (details below).

To fix this, this PR changes refresh listeners to be a noop when the shard is not yet serving reads (implicitly covering the recovery period). It doesn't matter anyway. 

There is a still a small problem I want to think how to solve - an indexing operation that came in with wait_for_refresh after the finalize recovery and before markAsDone is called will not be immediately visible when moving to `POST_RECOVERY`. I'm going to give it some more thought (I hope a simple refresh will do) but I think we can start reviewing on the main issue.

Deadlock with recovery:

When finalizing a peer recovery we mark the peer as "in sync". To do so we wait until the peer's local checkpoint is at least as high as the global checkpoint. If an operation with `refresh=wait_for` is added as a listener on that peer during recovery, it is not completed from the perspective of the primary. The primary than may wait for it to complete before advancing the local checkpoint for that peer. Since that peer is not considered in sync, the global checkpoint on the primary can be higher, causing a deadlock. Operation waits for recovery to finish and a refresh to happen. Recovery waits on the operation.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 1]
```diff
@@ -848,7 +848,7 @@ public long getWritingBytes() {
 
     public RefreshStats refreshStats() {
         // Null refreshListeners means this shard doesn't support them so there can't be any.
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: Drop the comment too?

File: core/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java [hunk 2]
```diff
@@ -1978,6 +1978,58 @@ public long indexTranslogOperations(List<Translog.Operation> operations, int tot
...
+        Settings settings = Settings.builder().put(IndexMetaData.SETTING_VERSION_CREATED, Version.CURRENT)
+            .put(IndexMetaData.SETTING_NUMBER_OF_REPLICAS, 1)
+            .put(IndexMetaData.SETTING_NUMBER_OF_SHARDS, 1)
+            .build();
+        IndexMetaData metaData = IndexMetaData.builder("test")
+            .putMapping("test", "{ \"properties\": { \"foo\":  { \"type\": \"text\"}}}")
+            .settings(settings)
+            .primaryTerm(0, 1).build();
+        IndexShard primary = newShard(new ShardId(metaData.getIndex(), 0), true, "n1", metaData, null);
+        recoveryShardFromStore(primary);
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java]: While we're here, can you fix the name of this method to be `recoverShardFromStore`?

**Review by jasontedor — APPROVED** [review]
LGTM.

Can you also revert f154e53b37e72d62c6eb429a5a5feeea86d40b5a and e1ef3d5cc270600c31159bf4d46eb180cb952f54?

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java (hunk 1 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: yep

File: core/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java (hunk 2 above)
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java]: fixed.

**Comment by bleskes** [comment]
@jasontedor thanks. Pushed another commit with a solution for the visibility issue. Can you take another look?

**Review by jasontedor — APPROVED** [review]
Visibility change looks good, so still LGTM.

**Pull request merged**